units/src/amount/mod.rs

```diff
@@ -38,6 +38,48 @@ pub use self::{
     unsigned::Amount,
 };
 
+/// Result of an operation on [`Amount`] or [`SignedAmount`].
+///
+/// The type parameter `T` should be normally `Amout` or `SignedAmount`.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum AmountOpResult<T> {
+    /// Result of a successful mathematical operation.
+    Valid(T),
+    /// Result of an unsuccessful mathematical operation.
+    Error(AmountOpError),
+}
+
+impl<T> AmountOpResult<T> {
+    /// Returns the contained valid amount.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the result is an `Error`.
+    pub fn unwrap_valid(self) -> T {
+        use AmountOpResult::*;
+
+        match self {
+            Valid(amount) => amount,
+            Error(_) => panic!("TODO"),
+        }
+    }
+
+    /// Converts this `AmountOpResult` to a `Result<T, AmountOpError>`.
+    pub fn valid(self) -> Result<T, AmountOpError> {
+        use AmountOpResult::*;
+
+        match self {
+            Valid(amount) => Ok(amount),
+            Error(e) => Err(e)
+        }
+    }
```

Comment:
I suspect having `and_then` method will be handy in our code and if we add it then we might as well make it `pub`.

Reply:
oh woops I did not action this yesterday. My bad

Reply:
Now done.